xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -89,337 +87,32 @@ pmap_registry_get(xlator_t *this)
- * the reason for the former is to ensure we don't end up with stale ports
- */
-int
-pmap_assign_port(xlator_t *this, int old_port, const char *path)
-{
-    int ret = -1;
-    int new_port = 0;
-
-    if (old_port) {
-        ret = pmap_registry_remove(this, 0, path, GF_PMAP_PORT_BRICKSERVER,
-                                   NULL, _gf_false);
-        if (ret) {
-            gf_msg(this->name, GF_LOG_WARNING, GD_MSG_PMAP_REGISTRY_REMOVE_FAIL,
-                   0,
-                   "Failed to"
-                   "remove pmap registry for older signin for path"
-                   " %s",
-                   path);
-        }
-    }
-    new_port = pmap_registry_alloc(this);
-    return new_port;
-}
-
-int
-pmap_registry_bind(xlator_t *this, int port, const char *brickname,
-                   gf_pmap_port_type_t type, void *xprt)
-{
-    struct pmap_registry *pmap = NULL;
-    int p = 0;
-
-    pmap = pmap_registry_get(this);
-
-    if (port > pmap->max_port)
-        goto out;
-
-    p = port;
-    if (pmap->ports[p].type == GF_PMAP_PORT_FREE) {
-        /* Because of some crazy race in volume start code path because
-         * of friend handshaking with volumes with quorum enabled we
-         * might end up into a situation where glusterd would start a
-         * brick and get a disconnect and then immediately try to start
-         * the same brick instance based on another friend update
-         * request. And then if for the very first brick even if the
-         * process doesn't come up at the end sign in event gets sent
-         * and we end up having two duplicate portmap entries for the
-         * same brick. Since in brick start we mark the previous port as
-         * free, its better to consider a sign in request as no op if
-         * the corresponding port type is marked as free
-         */
-        goto out;
-    }
-    if (pmap->ports[p].brickname) {
-        char *tmp = pmap->ports[p].brickname;
-        asprintf(&pmap->ports[p].brickname, "%s %s", tmp, brickname);
-        free(tmp);
-    } else {
-        pmap->ports[p].brickname = strdup(brickname);
-    }
-    pmap->ports[p].type = type;
-    pmap->ports[p].xprt = xprt;
-
-    gf_msg("pmap", GF_LOG_INFO, 0, GD_MSG_BRICK_ADD,
-           "adding brick %s on port %d", brickname, port);
-
-    if (pmap->last_alloc < p)
-        pmap->last_alloc = p;
-out:
-    return 0;
-}
-
-int
-pmap_registry_extend(xlator_t *this, int port, const char *brickname)
-{
-    struct pmap_registry *pmap = NULL;
-    char *old_bn;
-    char *new_bn;
-    size_t bn_len;
-    char *entry;
-    int found = 0;
-
-    pmap = pmap_registry_get(this);
-
-    if (port > pmap->max_port) {
-        return -1;
-    }
-
-    switch (pmap->ports[port].type) {
-        case GF_PMAP_PORT_LEASED:
-        case GF_PMAP_PORT_BRICKSERVER:
+    while (true) {
+        p = (rand() % (pmap->max_port - pmap->base_port + 1)) + pmap->base_port;
+        if (pmap_port_isfree(p)) {
```

Comment:
So, here we check if it's free by binding, then *closing* that bind. Why? Keep it open, use it!
(however, note that we deal only with IPv4 in pmap_port_isfree() , which may not be a good overall solution for IPv6 only setups).

I'd fix it in a follow-up patch, though.

Reply:
The reason for closing it here is because glusterd is not the process which actually wants to bind to it. Here, glusterd is doing the probe on behalf of the gluster brick process, selecting the port and then starting the brick process with the port number as an argument.

@nik-redhat If you are interested in further cleanups, this is certainly a good one in the same code area. Introduce a method where glusterd starts the brick process with a magic port number or no number. In that case, it will be up to the brick process to select the port to bind to. You will have to pass the range of valid ports(base port and max port) though. As you have already updated the pmap_registry in this PR, I don't think any change will be required once the port is selected by the brick process and it binds to it.

Reply:
So, the magic port here would be random port only, right? It too finds the port within the bas_port and max_port.
And, in the changes to something similar is done. Please pardon me if I understood something wrong. How else shall we do so then?

Reply:
I created an issue #1784 that has details on the other refactor I suggested above. We can discuss that change there. This PR looks good for what it is aiming for.

Reply:
Cool 👍🏼